website/blog/2018-08-24-7.0.0.md

```diff
@@ -0,0 +1,442 @@
+
+An example of this is with compiling a `class` (without `loose` mode on): 
+
+The specification says that you need to call a class with `new Person()` but if it's compiled to a function you could technically just do `Person()` so we include a runtime check for this.
+
+```js
+class Person {}
+```
+
+```js
+function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }
+
+var Person = function Person() {
+  _classCallCheck(this, Person);
+};
+```
+
+With `@babel/plugin-transform-runtime` and `@babel/runtime` (as a dependency), Babel can extract the individual functions and just require the modular functions to enable smaller output like so:
+
+```js
+var _classCallCheck = require("@babel/runtime/helpers/classCallCheck");
+
+var Person = function Person() {
+  _classCallCheck(this, Person);
+};
+```
+
+The same can be done with `external-helpers` and `rollup-plugin-babel`. We are looking into seeing if we can automatically do this in the future. Look out for a standalone post on Babel's helpers soon.
+
+## Automatic Polyfilling
+
+Polyfills are necessary for enabling features like `Promise`, `Symbol`, if supported browsers don't enable them. This is important when differentiating between what Babel does as a compiler (transforms *syntax*) vs. a polyfill (implements built-in functions/objects).
+
+It's easy enough to just import something that covers everything like `@babel/polyfill`:
+
+```js
+import "@babel/polyfill";
+```
+
+But it includes the whole polyfill, and you may not need to import everything if browsers support it already. This is the same problem that `@babel/preset-env` tries to solve with syntax, so we apply it here with polyfills. The option `useBuiltins: "entry"` does this by splitting the original import into the imports that are necessary.
+
+But we can do better by only importing the polyfills that are used in the codebase. The option `"useBuiltIns: "usage"` is our first attempt at enabling something like that ([docs](https://babeljs.io/docs/en/next/babel-preset-env#usebuiltins)).
+
+It will run through each file and inject an import at the top of each file if that built-in is "used" in the code. For example:
+
+```js
+import "core-js/modules/es6.promise";
+var a = new Promise();
+```
+
+The inference isn't perfect so there may be false positives.
+
+```js
+import "core-js/modules/es7.array.includes";
+a.includes // assume a is an []
+```
+
+Other ideas in this space are to use [polyfill.io](https://polyfill.io) if you are ok with relying on a service (or read how [Kent C. Dodds](https://twitter.com/kentcdodds) used that to [build a hosted service at PayPal](https://blog.kentcdodds.com/polyfill-as-needed-with-polyfill-service-35f0ff306a26)).
+
+## Misc
+
+### Babel Macros 🎣
```

Comment:
Why do we talk about `babel-plugin-macros` here? That looks like advertisement to me.

Reply:
Yeah I guess it's an advertisement in a way. It's written by Kent but it's pretty relevant to Babel: it seems like a lot of people are using it so I wanted to add a section to explain it? https://babeljs.io/blog/2017/09/11/zero-config-with-babel-macros we already did a blog post about it earlier. Just mentioning things that have happened since 7.0 (and previous blog posts if people didn't know about them, what do you think we should change here, just remove it?)

Reply:
yeah, I'm sorry for Kent, but it seems a bit unfair to promote his plugin here. Maybe wait for a second review.

Reply:
We don't have to do it, but I suggested it. It's seems to be more than just a single plugin since it has it's own ecosystem of "macros" so I thought it would be helpful.